### PR TITLE
[DI] Reduce complexity of autowiring

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no (tweaking existing ones)
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

- optional args are autowired
- methods with only optional args are autowired
- use default values of required args when possible

Both first changes remove the behavior delta between constructors and setters, and is expected to me now that things are more explicit. This reduces the "know-how" requirements for using autowiring and is easier to get correct intuitively. The 3rd change plays nice with named args.

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php [hunk 1]
```diff
@@ -243,44 +227,37 @@ private function autowireCalls(\ReflectionClass $reflectionClass, array $methodC
...
             $type = ProxyHelper::getTypeHint($reflectionMethod, $parameter, true);
 
             if (!$type) {
+                if (isset($arguments[$index])) {
+                    continue;
+                }
+
                 // no default value? Then fail
-                if (!$parameter->isOptional()) {
+                if (!$parameter->isDefaultValueAvailable()) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: as spotted here: https://github.com/symfony/symfony/pull/22254#discussion_r109444351

**Review by stof — APPROVED** [review]
👍

**Review by dunglas — APPROVED** [review]
👍

**Comment by fabpot** [comment]
@nicolas-grekas I let you merge this one

**Pull request merged**